### PR TITLE
Support debian/ubuntu installation of DPDK

### PR DESCRIPTION
**Pull request by szokovacs**
use the provided pkg-config if available

### CODE REVIEW

File: mk/PcapPlusPlus.mk.dpdk [hunk 1]
```diff
@@ -1,7 +1,15 @@
 ### DPDK ###
 
+PKGCONF = pkg-config
+
 # includes
+TMPI = $(shell $(PKGCONF) --cflags libdpdk)
```
Comment (seladb) [inline mk/PcapPlusPlus.mk.dpdk:6]: 1. If `libdpdk` isn't installed `pkg-config` gives an error:
    ```shell
    Package libdpdk was not found in the pkg-config search path.
    Perhaps you should add the directory containing `libdpdk.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'libdpdk' found
    ```
    This error will be prompt during the Makefile run and might be confusing to users. The only way I found to suppress it is adding `2> /dev/null`: `TMPI = $(shell $(PKGCONF) --cflags libdpdk 2> /dev/null)`

2.  Using `--cflags` gives more than just includes. Maybe we should consider using `--cflags-only-I` for the includes and then do:
    ```makefile
    TMPF = $(shell $(PKGCONF) --cflags-only-other libdpdk 2> /dev/null)
    PCAPPP_BUILD_FLAGS += $(TMPF)
    ```

File: mk/PcapPlusPlus.mk.dpdk [hunk 2]
```diff
@@ -10,4 +18,9 @@ PCAPPP_LIBS_DIR += -L$(RTE_SDK)/build/lib -L/lib64
 PCAPPP_BUILD_FLAGS += -msse -msse2 -msse3 -Wall
 
 # libs
+TMPL = $(shell $(PKGCONF) --libs libdpdk)
```
Comment (seladb) [inline mk/PcapPlusPlus.mk.dpdk:21]: Same here:
1. We can add `2> /dev/null` to suppress errors
2. We can separate into 2 parts:
    ```
    TMPLDIR = $(shell $(PKGCONF) --libs-only-L libdpdk)
    TMPL = $(shell $(PKGCONF) --libs-only-other libdpdk)
    ```

File: Pcap++/Makefile [hunk 3]
```diff
@@ -46,7 +46,9 @@ ifdef PF_RING_HOME
 INCLUDES += -I$(PF_RING_HOME)/userland/lib -I$(PF_RING_HOME)/kernel
 endif
 ifdef USE_DPDK
+PKGCONF = pkg-config
 INCLUDES += -I"$(RTE_SDK)/build/include"
+INCLUDES += $(shell $(PKGCONF) --cflags libdpdk)
```
Comment (seladb) [inline Pcap++/Makefile:51]: Why is this needed here if we already do it in `PcapPlusPlus.mk.dpdk`?

Comment (szokovacs) [reply Pcap++/Makefile:51]: I did not investigate too deeply, but on is INCLUDES, the other is PCAPPP_INCLUDES

File: configure-linux.sh [hunk 4]
```diff
@@ -274,13 +274,21 @@ fi
...
 function get_dpdk_version() {
    if [ -f $DPDK_HOME/pkg/dpdk.spec ]; then
       echo $(grep "Version" $DPDK_HOME/pkg/dpdk.spec | cut -d' ' -f2 | cut -d'.' -f 1,2);
-   else
+   elif [ -f $DPDK_HOME/VERSION ]; then
       echo $(cat $DPDK_HOME/VERSION | cut -d'.' -f 1,2);
+   elif [ -f $DPDK_HOME/rte_build_config.h ]; then
+      YEAR=$(grep RTE_VER_YEAR $DPDK_HOME/rte_build_config.h|cut -d ' ' -f3);
+      MONTH=$(grep RTE_VER_MONTH $DPDK_HOME/rte_build_config.h|cut -d ' ' -f3);
+      MINOR=$(grep RTE_VER_MINOR $DPDK_HOME/rte_build_config.h|cut -d ' ' -f3);
```
[the remark this answers is not in the document]
Comment (szokovacs) [reply configure-linux.sh:288]: I meant it as a last fallback, also I think I confused $DPDK_HOME/build with $DPDK_HOME/, I will fix it.

[the remark this answers is not in the document]
Comment (szokovacs) [reply configure-linux.sh:288]: Well, fixing it causes problems with my ubuntu-released 19.11.1, /usr/bin/ld: cannot find -lrte_pmd_vmxnet3_uio

File: Common++/Makefile [hunk 5]
```diff
@@ -40,7 +40,7 @@ endif
 			
 Obj/%.o: src/%.cpp Obj/%.d
 	@echo Building file: $<
-	@$(CXX) $(DEPS) $(INCLUDES) $(PCAPPP_BUILD_FLAGS) -c -fmessage-length=0 -MMD -MP -MF "$(@:Obj/%.o=Obj/%.d)" -MT "$@" -o "$@" "$<"
+	@$(CXX) $(DEPS) $(INCLUDES) $(PCAPPP_INCLUDES) $(PCAPPP_BUILD_FLAGS) -c -fmessage-length=0 -MMD -MP -MF "$(@:Obj/%.o=Obj/%.d)" -MT "$@" -o "$@" "$<"
```
Comment (seladb) [inline Common++/Makefile:43]: This change is not needed, `Common++` - `$(INCLUDES)` already has all the includes needed to compile this library

File: Packet++/Makefile [hunk 6]
```diff
@@ -40,7 +40,7 @@ endif
 
 Obj/%.o: src/%.cpp Obj/%.d
 	@echo Building file: $<
-	@$(CXX) $(DEPS) $(INCLUDES) $(PCAPPP_BUILD_FLAGS) -c -fmessage-length=0 -MMD -MP -MF "$(@:Obj/%.o=Obj/%.d)" -MT "$@" -o "$@" "$<"
+	@$(CXX) $(DEPS) $(INCLUDES) $(PCAPPP_INCLUDES) $(PCAPPP_BUILD_FLAGS) -c -fmessage-length=0 -MMD -MP -MF "$(@:Obj/%.o=Obj/%.d)" -MT "$@" -o "$@" "$<"
```
Comment (seladb) [inline Packet++/Makefile:43]: This change is not needed, `Packet++` - `$(INCLUDES)` already has all the includes needed to compile this library

File: Pcap++/Makefile [hunk 7]
```diff
@@ -46,7 +46,12 @@ ifdef PF_RING_HOME
 INCLUDES += -I$(PF_RING_HOME)/userland/lib -I$(PF_RING_HOME)/kernel
 endif
 ifdef USE_DPDK
+PKGCONF = pkg-config
+TMPI = $(shell $(PKGCONF) --cflags-only-I libdpdk 2>/dev/null)
 INCLUDES += -I"$(RTE_SDK)/build/include"
+ifneq ($(TMPI),)
+INCLUDES += $(TMPI)
+endif
```
Comment (seladb) [inline Pcap++/Makefile:54]: Now I remember - `$(INCLUDES)` should have all the header files needed to compile `Pcap++`. So this change is correct. Just one thing we can improve:
```Makefile
ifdef USE_DPDK
PKGCONF = pkg-config
TMPI = $(shell $(PKGCONF) --cflags-only-I libdpdk 2>/dev/null)
ifneq ($(TMPI),)
INCLUDES += $(TMPI)
else
INCLUDES += -I"$(RTE_SDK)/build/include"
endif
endif
```

File: mk/PcapPlusPlus.mk.dpdk [hunk 8]
```diff
@@ -1,13 +1,30 @@
...
 PCAPPP_LIBS_DIR += -L$(RTE_SDK)/build/lib -L/lib64
 
 #flags
 PCAPPP_BUILD_FLAGS += -msse -msse2 -msse3 -Wall
+ifneq ($(TMPFLAG),)
+PCAPPP_BUILD_FLAGS += $(TMPFLAG)
+endif
 
 # libs
+TMPL = $(shell $(PKGCONF) --libs libdpdk 2>/dev/null)
```
Comment (seladb) [inline mk/PcapPlusPlus.mk.dpdk:25]: here we should use the `--libs-only-l` flag

File: mk/PcapPlusPlus.mk.dpdk [hunk 9]
```diff
@@ -1,13 +1,30 @@
...
+TMPFLAG = $(shell $(PKGCONF) --cflags-only-other libdpdk 2>/dev/null)
+
+ifeq ($(TMPI),)
 PCAPPP_INCLUDES += -I$(RTE_SDK)/build/include
+else
+PCAPPP_INCLUDES += $(TMPI)
+endif
 
 # libs dir
 PCAPPP_LIBS_DIR += -L$(RTE_SDK)/build/lib -L/lib64
```
Comment (seladb) [inline mk/PcapPlusPlus.mk.dpdk:16]: Here we can do a similar thing with the `--libs-only-L` flag:
```Makefile
TMPLD = $(shell $(PKGCONF) --libs-only-L libdpdk 2>/dev/null)
ifneq ($(TMPLD),)
PCAPPP_LIBS_DIR+= $(TMPLD)
else
PCAPPP_LIBS_DIR += -L$(RTE_SDK)/build/lib -L/lib64
endif
```

File: configure-linux.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (seladb) [reply configure-linux.sh]: ok, we can keep it like this since it's the 3rd fallback option

File: Pcap++/Makefile (hunk 3 above)
Comment (seladb) [reply Pcap++/Makefile]: You're right, we should use `INCLUDES` and not `PCAPPP_INCLUDES`, please see my comment: https://github.com/seladb/PcapPlusPlus/pull/797#discussion_r799228388

File: Common++/Makefile (hunk 5 above)
Comment (szokovacs) [reply Common++/Makefile:43]: I agree, it should not be, but --cflags-only-other is for me "-include rte_config.h -march=corei7" and we included it in $(PCAPPP_BUILD_FLAGS) in commit b9f00a9c85b312e9fe84a8b4093edc3985dd425d

Comment (seladb) [reply Common++/Makefile:43]: I'm not sure I understand... 🤔

Comment (szokovacs) [reply Common++/Makefile:43]: $ pkg-config --cflags-only-other libdpdk
-include rte_config.h -march=corei7

in mk/PcapPlusPlus.mk.dpdk :
TMPFLAG = $(shell $(PKGCONF) --cflags-only-other libdpdk 2>/dev/null)
ifneq ($(TMPFLAG),)
PCAPPP_BUILD_FLAGS += $(TMPFLAG)
endif

Comment (seladb) [reply Common++/Makefile:43]: I'm not sure how this is related to the `Makefile` in `Common++`...
Basically `Common++` and `Packet++` libraries don't use any DPDK object, so they shouldn't need any additional DPDK include. Only the `Pcap++` library needs DPDK header files, which is why the change you made is necessary

File: mk/PcapPlusPlus.mk.dpdk [hunk 10]
```diff
@@ -1,13 +1,35 @@
...
+TMPFLAG = $(shell $(PKGCONF) --cflags-only-other libdpdk 2>/dev/null)
+
+ifeq ($(TMPI),)
 PCAPPP_INCLUDES += -I$(RTE_SDK)/build/include
+else
+PCAPPP_INCLUDES += $(TMPI)
+endif
 
 # libs dir
+TMPLD = $(shell $(PKGCONF) --libs-l libdpdk 2>/dev/null)
```
Comment (seladb) [inline mk/PcapPlusPlus.mk.dpdk:16]: I'm not sure the `--libs-l` flag exists, isn't it `--libs-only-l`?

Comment (szokovacs) [reply mk/PcapPlusPlus.mk.dpdk]: Sorry about this, the 2>/dev/null hid the typo :(

File: Common++/Makefile (hunk 5 above)
Comment (szokovacs) [reply Common++/Makefile:43]: Common++/Makefile includes mk/PcapPlusPlus.mk which in turn has mk/PcapPlusPlus.mk.dpdk included, so PCAPPP_BUILD_FLAGS ends up including the DPDK stuff. $(INCLUDES) $(PCAPPP_BUILD_FLAGS) seems asymmetric to me, maybe there should be a BUILD_FLAGS too?
Or should the PcapPlusPlus.mk.dpdk restricted to Pcap++ ?

Comment (seladb) [reply Common++/Makefile:43]: `PCAPPP_INCLUDES`, `PCAPPP_LIBS`, `PCAPPP_LIBS_DIR` and  `PCAPPP_BUILD_FLAGS` are mostly for external users who want to use PcapPlusPlus.
PcapPlusPlus own libraries (`Common++`, `Packet++`, `Pcap++`) have their own includes so they don't need to user `PCAPPP_*`

File: Common++/Makefile [hunk 11]
```diff
@@ -40,7 +40,7 @@ endif
 			
 Obj/%.o: src/%.cpp Obj/%.d
 	@echo Building file: $<
-	@$(CXX) $(DEPS) $(INCLUDES) $(PCAPPP_BUILD_FLAGS) -c -fmessage-length=0 -MMD -MP -MF "$(@:Obj/%.o=Obj/%.d)" -MT "$@" -o "$@" "$<"
+	@$(CXX) $(DEPS) $(INCLUDES) $(BUILD_FLAGS) -c -fmessage-length=0 -MMD -MP -MF "$(@:Obj/%.o=Obj/%.d)" -MT "$@" -o "$@" "$<"
```
Comment (seladb) [inline Common++/Makefile:43]: Let's just revert the change and keep it as it was?

File: Packet++/Makefile [hunk 12]
```diff
@@ -40,7 +40,7 @@ endif
 
 Obj/%.o: src/%.cpp Obj/%.d
 	@echo Building file: $<
-	@$(CXX) $(DEPS) $(INCLUDES) $(PCAPPP_BUILD_FLAGS) -c -fmessage-length=0 -MMD -MP -MF "$(@:Obj/%.o=Obj/%.d)" -MT "$@" -o "$@" "$<"
+	@$(CXX) $(DEPS) $(INCLUDES) $(BUILD_FLAGS) -c -fmessage-length=0 -MMD -MP -MF "$(@:Obj/%.o=Obj/%.d)" -MT "$@" -o "$@" "$<"
```
Comment (seladb) [inline Packet++/Makefile:43]: Let's just revert the change and keep it as it was?

File: mk/PcapPlusPlus.mk.common [hunk 13]
```diff
@@ -24,3 +24,5 @@ CXXFLAGS := -O2 -g -Wall
 endif
 
 PCAPPP_BUILD_FLAGS += $(CXXFLAGS)
+
+BUILD_FLAGS := $(PCAPPP_BUILD_FLAGS)
```
Comment (seladb) [inline mk/PcapPlusPlus.mk.common:28]: Let's just revert the change and keep it as it was?

Comment (szokovacs) [reply mk/PcapPlusPlus.mk.common:28]: Which version do you mean? Keep PCAPPP_BUILD_FLAGS as it was and extend PCAPPP_INCLUDES with the output of
"pkg-config --cflags" ?

Comment (seladb) [reply mk/PcapPlusPlus.mk.common:28]: We can remove `BUILD_FLAGS := $(PCAPPP_BUILD_FLAGS)` from this file and use `$(PCAPPP_BUILD_FLAGS)` in `Common++/Makefile` and `Packet++/Makefile` (which is the same as we currently have in `dev` and `master` branches).

I suggest we focus this PR on making DPDK build work with `pkg-config`. We can leave additional improvements to a separate PR. Please let me know what you think

File: configure-linux.sh [hunk 14]
```diff
@@ -300,6 +308,7 @@ if (( $COMPILE_WITH_DPDK > 0 )) ; then
 
    # if DPDK ver < 20.11 concat additional definitions to PcapPlusPlus.mk
    CUR_DPDK_VERSION=$(get_dpdk_version)
+   echo $CUR_DPDK_VERSION
```
Comment (seladb) [inline configure-linux.sh:311]: Is this a leftover from your debugging sessions? I'm not sure we need this print 🤔

Comment (szokovacs) [reply configure-linux.sh:311]: Yes, sorry.

File: mk/PcapPlusPlus.mk.common (hunk 13 above)
Comment (szokovacs) [reply mk/PcapPlusPlus.mk.common]: I agree, pushed a new version

**Review by seladb — APPROVED** [review]

**Comment by seladb** [comment]
Thanks for your contribution @szokovacs ! I really appreciate your hard work and I'm sorry for the time it took to review and merge this PR

**Comment by szokovacs** [comment]
> Thanks for your contribution @szokovacs ! I really appreciate your hard work and I'm sorry for the time it took to review and merge this PR

No problem at all, I agree it's important to iron out a details :)
